### ISSUE TITLE
tui: segfault on resize

### ISSUE THREAD

**Issue by oatmealraisin** [neovim/neovim#7572]
```
NVIM v0.2.2-38-gf8d40e7d5
Build type: Release
LuaJIT 2.0.5
Compilation: /usr/bin/cc -Wconversion -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -O2 -DNDEBUG -DMIN_LOG_LEVEL=3 -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -I/home/ryan/documents/sources/neovim/build/config -I/home/ryan/documents/sources/neovim/src -I/home/ryan/documents/sources/neovim/.deps/usr/include -I/home/ryan/documents/sources/neovim/.deps/usr/include -I/home/ryan/documents/sources/neovim/.deps/usr/include -I/home/ryan/documents/sources/neovim/.deps/usr/include -I/home/ryan/documents/sources/neovim/.deps/usr/include -I/home/ryan/documents/sources/neovim/.deps/usr/include -I/usr/include -I/home/ryan/documents/sources/neovim/build/src/nvim/auto -I/home/ryan/documents/sources/neovim/build/include
Compiled by ryan@cortado
Features: +acl +iconv +jemalloc +tui 
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/local/share/nvim"

Run :checkhealth for more info
```

OS:
```
Linux cortado 4.13.12-1-ARCH #1 SMP PREEMPT Wed Nov 8 11:54:06 CET 2017 x86_64 GNU/Linux
```

- Terminal name/version:
```
 $ xterm -v
XTerm(330)
```
- `$TERM`:
```
xterm-256color
```

### Steps to reproduce using `nvim -u NORC`

```
nvim -u NORC -V9logs
```

### Actual behaviour

Screen flashes, input lags, and segfault occurs on `:q!`

### Expected behaviour

Unnamed buffer opens

**Comment by oatmealraisin** [comment]
Here are the logs
[Logs](https://gist.github.com/241dfb2471a1701bd42a1e48392c4724)

**Comment by oatmealraisin** [comment]
Interestingly, when -u NORC is not specified, it takes until a screen resize or backgrounding to get to the state I described

**Comment by justinmk** [comment]
Can you please post a backtrace using the steps from the "building Neovim" wiki page?

**Comment by justinmk** [comment]
Correction, here is the page https://github.com/neovim/neovim/wiki/Development-tips#backtrace-linux

**Comment by oatmealraisin** [comment]
Not sure I understand. My ulimit for core file size is unlimited, but there is no core file. The segfault only happens when I exit vim, could that keep it from creating one?

**Comment by justinmk** [comment]
@oatmealraisin No. What makes you think it is a segfault? What is the exact error you see?

**Comment by oatmealraisin** [comment]
`zsh: segmentation fault (core dumped)  nvim -u NORC`

**Comment by justinmk** [comment]
> My ulimit for core file size is unlimited

You must run `ulimit -c unlimited` in the current shell. It doesn't persist. Please do that, then provoke the segfault, and it should produce a core somewhere. If the `core` file doesn't appear in the current directory, https://stackoverflow.com/a/18368068/152142 may help (or Arch linux wiki, I guess).

**Comment by oatmealraisin** [comment]
I was able to find the core dumps through coredumpctl, is this the output you're looking for?
```
(gdb) bt
#0  0x00007f645176f8a0 in raise () from /usr/lib/libc.so.6
#1  0x00007f6451770f09 in abort () from /usr/lib/libc.so.6
#2  0x000055cb2d56bcc7 in uv_mutex_lock (mutex=<optimized out>)
    at /home/ryan/documents/sources/neovim/.deps/build/src/libuv/src/unix/thread.c:129
#3  0x000055cb2d3fe228 in loop_schedule ()
#4  0x000055cb2d54e2cd in ui_bridge_mode_info_set ()
#5  0x000055cb2d54b0d9 in ui_call_mode_info_set ()
#6  0x000055cb2d54cd41 in ui_mode_info_set ()
#7  0x000055cb2d54ce89 in ui_refresh ()
#8  0x000055cb2d3feab9 in multiqueue_process_events ()
#9  0x000055cb2d54e3f5 in ui_bridge_stop ()
#10 0x000055cb2d54b93b in ui_builtin_stop ()
#11 0x000055cb2d4c2702 in mch_exit ()
#12 0x000055cb2d46a1c5 in getout ()
#13 0x000055cb2d418d28 in ex_quit ()
#14 0x000055cb2d41fcce in do_cmdline ()
#15 0x000055cb2d497c67 in nv_colon ()
#16 0x000055cb2d4a0f38 in normal_execute ()
#17 0x000055cb2d52e176 in state_enter ()
#18 0x000055cb2d4985e9 in normal_enter ()
#19 0x000055cb2d386dc6 in main ()
```

**Comment by justinmk** [comment]
@oatmealraisin great, thanks! Would be even better if you could follow the steps at https://github.com/neovim/neovim/wiki/Development-tips#backtrace-linux

tl;dr:

    gdb path/to/nvim path/to/core 2>&1 | tee backtrace.txt
    thread apply all bt full

then upload backtrace.txt to https://gist.github.com. I promise I'll stop pestering you after that :)

**Comment by oatmealraisin** [comment]
Unfortunately, the core files were formatted an odd way that gdb didn't like, sorry. [This](gdb -q -n -ex bt -batch ./build/bin/nvim <path/to/core> > backtrace.txt) is as close as I can find. The core dumps are being handled by systemd, these were found in /var/lib/systemd/coredump as an lz4 archive. After decompression, I fed these into the gdb call provided in the wiki. Hope this helps!

**Comment by oni-link** [comment]
@oatmealraisin, I could not reproduce(missing steps?) the segfault you described.
However running `nvim -u NONE` and resizing the terminal I got a segfault:
```
==15999==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x608004680cf8 at pc 0x000001842286 bp 0x7f0dd122ae90 sp 0x7f0dd122ae88
READ of size 8 at 0x608004680cf8 thread T1
    #0 0x1842285 in ugrid_put /home/oni-link/git/neovim/build/../src/nvim/ugrid.c:107:17
    #1 0x1817d1f in tui_put /home/oni-link/git/neovim/build/../src/nvim/tui/tui.c:979:10
    #2 0x186ab66 in ui_bridge_put_event /home/oni-link/git/neovim/build/src/nvim/auto/ui_events_bridge.generated.h:141:3
    #3 0xa32eba in multiqueue_process_events /home/oni-link/git/neovim/build/../src/nvim/event/multiqueue.c:150:7
    #4 0xa2ccf9 in loop_poll_events /home/oni-link/git/neovim/build/../src/nvim/event/loop.c:59:3
    #5 0x181b167 in tui_main /home/oni-link/git/neovim/build/../src/nvim/tui/tui.c:343:5
    #6 0x1866f00 in ui_thread_run /home/oni-link/git/neovim/build/../src/nvim/ui_bridge.c:104:3
    #7 0x51958e in __asan::AsanThread::ThreadStart(unsigned long, __sanitizer::atomic_uintptr_t*) /home/oni-link/git/llvm/projects/compiler-rt/lib/asan/asan_thread.cc:259
    #8 0x7f0dd5464616  (/lib64/libpthread.so.0+0x7616)
    #9 0x7f0dd4665bce in clone (/lib64/libc.so.6+0xf9bce)

0x608004680cf8 is located 0 bytes to the right of 88-byte region [0x608004680ca0,0x608004680cf8)
allocated by thread T1 here:
    #0 0x50baa0 in malloc /home/oni-link/git/llvm/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:88
    #1 0xf54101 in try_malloc /home/oni-link/git/neovim/build/../src/nvim/memory.c:87:15
    #2 0xf54329 in xmalloc /home/oni-link/git/neovim/build/../src/nvim/memory.c:121:15
    #3 0x183e35b in ugrid_resize /home/oni-link/git/neovim/build/../src/nvim/ugrid.c:32:17
    #4 0x1813098 in tui_resize /home/oni-link/git/neovim/build/../src/nvim/tui/tui.c:737:3
    #5 0x1867c48 in ui_bridge_resize_event /home/oni-link/git/neovim/build/src/nvim/auto/ui_events_bridge.generated.h:4:3
    #6 0xa32eba in multiqueue_process_events /home/oni-link/git/neovim/build/../src/nvim/event/multiqueue.c:150:7
    #7 0xa2ccf9 in loop_poll_events /home/oni-link/git/neovim/build/../src/nvim/event/loop.c:59:3
    #8 0x181b167 in tui_main /home/oni-link/git/neovim/build/../src/nvim/tui/tui.c:343:5
    #9 0x1866f00 in ui_thread_run /home/oni-link/git/neovim/build/../src/nvim/ui_bridge.c:104:3
    #10 0x51958e in __asan::AsanThread::ThreadStart(unsigned long, __sanitizer::atomic_uintptr_t*) /home/oni-link/git/llvm/projects/compiler-rt/lib/asan/asan_thread.cc:259

Thread T1 created by T0 here:
    #0 0x465ec0 in __interceptor_pthread_create /home/oni-link/git/llvm/projects/compiler-rt/lib/asan/asan_interceptors.cc:204
    #1 0x19331d0 in uv_thread_create /home/oni-link/git/neovim/.deps/build/src/libuv/src/unix/thread.c:75
    #2 0x1812be3 in tui_start /home/oni-link/git/neovim/build/../src/nvim/tui/tui.c:170:10
    #3 0x185274c in ui_builtin_start /home/oni-link/git/neovim/build/../src/nvim/ui.c:125:3
    #4 0xe47e2e in main /home/oni-link/git/neovim/build/../src/nvim/main.c:454:5
    #5 0x7f0dd458c439 in __libc_start_main (/lib64/libc.so.6+0x20439)

SUMMARY: AddressSanitizer: heap-buffer-overflow /home/oni-link/git/neovim/build/../src/nvim/ugrid.c:107:17 in ugrid_put
```
Looks like `grid->row` is increased too much. Can you test the following patch for the resizing problem?

```diff
diff --git a/src/nvim/tui/tui.c b/src/nvim/tui/tui.c
index 8e0e905bc..1a65a78cb 100644
--- a/src/nvim/tui/tui.c
+++ b/src/nvim/tui/tui.c
@@ -467,7 +467,7 @@ static void final_column_wrap(UI *ui)
   UGrid *grid = &data->grid;
   if (grid->col == ui->width) {
     grid->col = 0;
-    if (grid->row < ui->height) {
+    if (grid->row < MIN(ui->height, grid->height-1)) {
       grid->row++;
     }
   }

```
Your second backtrace could hint that `grid->cells` was deallocated twice.
Perhaps this patch helps a bit?

```diff
diff --git a/src/nvim/ugrid.c b/src/nvim/ugrid.c
index 2b5e96ee6..e0880b4c7 100644
--- a/src/nvim/ugrid.c
+++ b/src/nvim/ugrid.c
@@ -135,6 +135,7 @@ static void destroy_cells(UGrid *grid)
       xfree(grid->cells[i]);
     }
     xfree(grid->cells);
+    grid->cells = NULL;
   }
 }
```

@justinmk, I also noticed some memory leaks when the terminal is resized a lot and events are purged. Some events contain pointers to allocated memory that the event handler frees, but not in this situation. Problematic functions
in `ui_events_bridge.generated.h` contain calls to `copy_string()` or `copy_array()`:

```
Direct leak of 1988 byte(s) in 284 object(s) allocated from:
    #0 0x50baa0 in malloc /home/oni-link/git/llvm/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:88
    #1 0xf54101 in try_malloc /home/oni-link/git/neovim/build/../src/nvim/memory.c:87:15
    #2 0xf54329 in xmalloc /home/oni-link/git/neovim/build/../src/nvim/memory.c:121:15
    #3 0xf548e8 in xmallocz /home/oni-link/git/neovim/build/../src/nvim/memory.c:196:15
    #4 0xf54a6e in xmemdupz /home/oni-link/git/neovim/build/../src/nvim/memory.c:214:17
    #5 0x60f5b0 in copy_string /home/oni-link/git/neovim/build/../src/nvim/api/private/helpers.c:1010:30
    #6 0x186433c in ui_bridge_mode_change /home/oni-link/git/neovim/build/src/nvim/auto/ui_events_bridge.generated.h:114:22
    #7 0x1845e8d in ui_call_mode_change /home/oni-link/git/neovim/build/src/nvim/auto/ui_events_call.generated.h:53:3
    #8 0x185ab38 in ui_cursor_shape /home/oni-link/git/neovim/build/../src/nvim/ui.c:545:5
    #9 0x185a100 in ui_refresh /home/oni-link/git/neovim/build/../src/nvim/ui.c:292:3
    #10 0x185b3e0 in ui_refresh_event /home/oni-link/git/neovim/build/../src/nvim/ui.c:297:3
    #11 0xa32eba in multiqueue_process_events /home/oni-link/git/neovim/build/../src/nvim/event/multiqueue.c:150:7
    #12 0xa2ccf9 in loop_poll_events /home/oni-link/git/neovim/build/../src/nvim/event/loop.c:59:3
    #13 0x123a0d6 in input_poll /home/oni-link/git/neovim/build/../src/nvim/os/input.c:349:3
    #14 0x123587f in inbuf_poll /home/oni-link/git/neovim/build/../src/nvim/os/input.c:372:24
    #15 0x12356e5 in os_inchar /home/oni-link/git/neovim/build/../src/nvim/os/input.c:115:18
    #16 0x1703f4e in state_enter /home/oni-link/git/neovim/build/../src/nvim/state.c:53:13
    #17 0x10556b4 in normal_enter /home/oni-link/git/neovim/build/../src/nvim/normal.c:466:3
    #18 0xe48bf3 in main /home/oni-link/git/neovim/build/../src/nvim/main.c:569:3
    #19 0x7f542737f439 in __libc_start_main (/lib64/libc.so.6+0x20439)

Indirect leak of 2944512 byte(s) in 4828 object(s) allocated from:
    #0 0x50bf20 in realloc /home/oni-link/git/llvm/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:107
    #1 0xf546e1 in xrealloc /home/oni-link/git/neovim/build/../src/nvim/memory.c:169:15
    #2 0x60efdd in copy_object /home/oni-link/git/neovim/build/../src/nvim/api/private/helpers.c:1045:9
    #3 0x60ff93 in copy_array /home/oni-link/git/neovim/build/../src/nvim/api/private/helpers.c:1020:5
    #4 0x1862d97 in ui_bridge_mode_info_set /home/oni-link/git/neovim/build/src/nvim/auto/ui_events_bridge.generated.h:51:30
    #5 0x1844a86 in ui_call_mode_info_set /home/oni-link/git/neovim/build/src/nvim/auto/ui_events_call.generated.h:23:3
    #6 0x185a6a2 in ui_mode_info_set /home/oni-link/git/neovim/build/../src/nvim/ui.c:472:3
    #7 0x185a0f0 in ui_refresh /home/oni-link/git/neovim/build/../src/nvim/ui.c:290:3
    #8 0x185b3e0 in ui_refresh_event /home/oni-link/git/neovim/build/../src/nvim/ui.c:297:3
    #9 0xa32eba in multiqueue_process_events /home/oni-link/git/neovim/build/../src/nvim/event/multiqueue.c:150:7
    #10 0xa2ccf9 in loop_poll_events /home/oni-link/git/neovim/build/../src/nvim/event/loop.c:59:3
    #11 0x123a0d6 in input_poll /home/oni-link/git/neovim/build/../src/nvim/os/input.c:349:3
    #12 0x123587f in inbuf_poll /home/oni-link/git/neovim/build/../src/nvim/os/input.c:372:24
    #13 0x12356e5 in os_inchar /home/oni-link/git/neovim/build/../src/nvim/os/input.c:115:18
    #14 0x1703f4e in state_enter /home/oni-link/git/neovim/build/../src/nvim/state.c:53:13
    #15 0x10556b4 in normal_enter /home/oni-link/git/neovim/build/../src/nvim/normal.c:466:3
    #16 0xe48bf3 in main /home/oni-link/git/neovim/build/../src/nvim/main.c:569:3
    #17 0x7f542737f439 in __libc_start_main (/lib64/libc.so.6+0x20439)
```

**Comment by oatmealraisin** [comment]
@oni-link There are no directions, per se. I open neovim with -u NORC, and the bug immediately starts, and when I `:q!`, neovim closes with a segfault. I'm going to continue trying to get a proper core file. I'll also apply each patch you've given. Thank you so much for the help!

**Comment by oatmealraisin** [comment]
@oni-link @justinmk I should have tested this first, but the error isn't reproducable on terminator, or while using tmux. I'm running on Wayland, so I'm willing to chalk this up to Xwayland. Feel free to close this, or I can continue debugging for you. Thanks again for the help.

**Comment by justinmk** [comment]
@oni-link Thank you very much. Do you have time to make a test and send a PR?

> I also noticed some memory leaks when the terminal is resized a lot and events are purged. Some events contain pointers to allocated memory that the event handler frees, but not in this situation.

That doesn't surprise me, the "purge" is a big hack that is not supposed to happen except if things have gone really wrong. Any event handler that is supposed to free objects won't get executed. Not sure if there's anything we can do about this.

**Comment by oni-link** [comment]
> @oni-link Thank you very much. Do you have time to make a test and send a PR?

@justinmk, sorry no time. In my tests the crash happened when the terminal size was repeatedly reduced fast to one character.
